### ISSUE TITLE
Port K-Means and GMM clustering-based outlier detection from EDPIE

### ISSUE THREAD

**Issue by eberrigan** [talmolab/sleap-roots-analyze#8]
## Summary

Port K-Means and Gaussian Mixture Model (GMM) clustering-based outlier detection methods from the EDPIE_wheat_analysis repository. These methods complement existing outlier detection approaches and provide alternative perspectives on data structure.

## Background

**Current outlier detection methods in sleap-roots-analyze:**
- ✅ PCA reconstruction error
- ✅ Isolation Forest
- ✅ Mahalanobis distance with chi-squared thresholds (#7 adds GOF testing)
- ❌ K-Means clustering (exists in EDPIE_wheat_analysis)
- ❌ GMM clustering (exists in EDPIE_wheat_analysis)

**Why add clustering-based methods:**
1. Different perspective: Clustering identifies natural groupings, outliers are points far from all clusters
2. Interpretability: Cluster assignments provide biological insight (genotype groups, experimental batches)
3. Foundation for cluster-aware analyses: Enables future cluster-specific outlier detection, heritability, etc.
4. Already implemented: Proven code exists in EDPIE project

## Source Code Location

**EDPIE_wheat_analysis repository:**
- `src/outlier_detection.py`:
  - `detect_outliers_kmeans()` (lines ~336-450)
  - `detect_outliers_gmm()` (lines ~773-900)
- `src/visualization_outlier_methods.py`:
  - `create_kmeans_outlier_plots()` (lines ~84-250)
  - `create_gmm_outlier_plots()` (lines ~753-920)

### 1. Port K-Means Outlier Detection

**Add to `src/sleap_roots_analyze/outlier_detection.py`:**

```python
def detect_outliers_kmeans(
    data: Union[pd.DataFrame, np.ndarray],
    n_clusters: Optional[int] = 3,
    distance_threshold: float = 2.0,
    standardize: bool = True,
    random_state: int = 42
) -> Dict[str, Any]:
    """
    Detect outliers using K-means clustering.

    Samples far from all cluster centers are considered outliers.

    Args:
        data: Input data
        n_clusters: Number of clusters (auto-adjusted if too large)
        distance_threshold: Threshold in standard deviations
        standardize: Standardize data before clustering
        random_state: Random seed

    Returns:
        {
            'method': 'KMeans',
            'outlier_indices': List[int],
            'n_outliers': int,
            'n_clusters': int,
            'cluster_labels': np.ndarray,
            'cluster_sizes': List[int],
            'min_distances_to_centers': np.ndarray,
            'threshold_value': float,
            'inertia': float,  # Within-cluster sum of squares
            ...
        }
    """
```

**Enhancements while porting:**
- Standardize return dict structure to match other methods
- Add automatic cluster number selection (if n_clusters too large for data)
- Include cluster quality metrics (silhouette score, inertia)
- Add comprehensive error handling
- Update docstrings to match project style

### 2. Port GMM Outlier Detection

**Add to `src/sleap_roots_analyze/outlier_detection.py`:**

```python
def detect_outliers_gmm(
    data: Union[pd.DataFrame, np.ndarray],
    n_components: Optional[int] = None,
    max_components: int = 5,
    percentile_threshold: float = 99.0,
    covariance_type: str = 'full',
    standardize: bool = True,
    random_state: int = 42
) -> Dict[str, Any]:
    """
    Detect outliers using Gaussian Mixture Model.

    GMM models data as mixture of Gaussians. Outliers have low probability
    under the fitted model.

    Args:
        data: Input data
        n_components: Number of mixture components (auto-selected if None)
        max_components: Max components to test for auto-selection
        percentile_threshold: Percentile for outlier threshold (95-99.9)
        covariance_type: 'full', 'tied', 'diag', or 'spherical'
        standardize: Standardize data before GMM
        random_state: Random seed

    Returns:
        {
            'method': 'GMM',
            'outlier_indices': List[int],
            'n_outliers': int,
            'n_components': int,
            'cluster_labels': np.ndarray,  # Hard assignments
            'probabilities': np.ndarray,  # Soft assignments
            'log_likelihoods': np.ndarray,  # Per-sample log P(x)
            'bic': float,
            'aic': float,
            'bic_scores': List[float],  # For all k tested
            'aic_scores': List[float],
            'converged': bool,
            ...
        }
    """
```

**Enhancements while porting:**
- Add BIC/AIC scores for model selection transparency
- Include both hard (cluster labels) and soft (probabilities) assignments
- Add convergence diagnostics
- Return goodness-of-fit metrics

### 3. Port Visualization Functions

**Add to `src/sleap_roots_analyze/outlier_visualization.py`:**

```python
def create_kmeans_outlier_plots(
    df: pd.DataFrame,
    kmeans_results: Dict
) -> Dict[str, plt.Figure]:
    """
    Create K-Means specific outlier detection plots.

    Generates:
    1. PCA scatter with cluster coloring + outliers
    2. Distance distribution with threshold
    3. Cluster sizes bar chart
    """

def create_gmm_outlier_plots(
    df: pd.DataFrame,
    gmm_results: Dict
) -> Dict[str, plt.Figure]:
    """
    Create GMM specific outlier detection plots.

    Generates:
    1. PCA scatter with probability-based coloring
    2. Log-likelihood distribution with threshold
    3. BIC/AIC scores for component selection
    4. Soft cluster assignment visualization
    """
```

### 4. Update `combine_outlier_methods()`

**Modify `src/sleap_roots_analyze/outlier_detection.py`:**

```python
def combine_outlier_methods(
    pca_results: Optional[Dict] = None,
    isolation_results: Optional[Dict] = None,
    mahalanobis_results: Optional[Dict] = None,
    kmeans_results: Optional[Dict] = None,  # NEW
    gmm_results: Optional[Dict] = None,     # NEW
    consensus_threshold: float = 0.5
) -> Dict:
    """Combine outlier detection results from multiple methods."""
```

### Unit Tests (`tests/test_outlier_detection.py`)

```python
def test_detect_outliers_kmeans():
    """Test K-Means outlier detection with known data."""
    # Test with synthetic data with clear outliers
    # Test automatic cluster number adjustment
    # Test with various thresholds
    # Test error handling (empty data, NaN, etc.)

def test_detect_outliers_gmm():
    """Test GMM outlier detection."""
    # Test with synthetic multi-modal data
    # Test automatic component selection via BIC
    # Test different covariance types
    # Test convergence

def test_kmeans_gmm_consistency():
    """Test that K-Means and GMM give reasonable results."""
    # On same data, should identify similar outliers
```

### Visualization Tests (`tests/test_outlier_visualization.py`)

```python
def test_create_kmeans_outlier_plots():
    """Test K-Means visualization functions."""

def test_create_gmm_outlier_plots():
    """Test GMM visualization functions."""
```

### Integration Test

- Run all 5 methods on wheat phenotyping dataset (925 samples, 34 traits)
- Compare outlier overlap across methods
- Verify visualizations generate correctly
- Check performance (should complete in <30 seconds)

## Files to Modify

- `src/sleap_roots_analyze/outlier_detection.py`
  - Add `detect_outliers_kmeans()`
  - Add `detect_outliers_gmm()`
  - Update `combine_outlier_methods()`

- `src/sleap_roots_analyze/outlier_visualization.py`
  - Add `create_kmeans_outlier_plots()`
  - Add `create_gmm_outlier_plots()`

- `tests/test_outlier_detection.py`
  - Add K-Means tests
  - Add GMM tests

- `tests/test_outlier_visualization.py`
  - Add visualization tests

## Dependencies

All already available in `scikit-learn`:
- `sklearn.cluster.KMeans` ✓
- `sklearn.mixture.GaussianMixture` ✓

No new dependencies required!

## Example Usage

```python
from sleap_roots_analyze.outlier_detection import (
    detect_outliers_kmeans,
    detect_outliers_gmm,
    combine_outlier_methods
)
from sleap_roots_analyze.outlier_visualization import (
    create_kmeans_outlier_plots,
    create_gmm_outlier_plots
)

# Get numeric traits
numeric_traits = df[trait_cols].dropna()

# Run clustering-based outlier detection
kmeans_results = detect_outliers_kmeans(
    numeric_traits,
    n_clusters=3,
    distance_threshold=2.0
)

gmm_results = detect_outliers_gmm(
    numeric_traits,
    n_components=None,  # Auto-select
    percentile_threshold=99.0
)

# Combine with other methods
combined = combine_outlier_methods(
    pca_results=pca_results,
    isolation_results=iso_results,
    mahalanobis_results=mahal_results,
    kmeans_results=kmeans_results,
    gmm_results=gmm_results,
    consensus_threshold=0.5  # 50% agreement
)

# Visualize
kmeans_figs = create_kmeans_outlier_plots(df, kmeans_results)
gmm_figs = create_gmm_outlier_plots(df, gmm_results)
```

## Benefits

1. **More detection methods**: 5 total methods for comprehensive outlier detection
2. **Clustering insights**: Learn about data structure (genotype groups, batches)
3. **Foundation for future work**: Enables cluster-aware analyses (Issues #9-11)
4. **Proven code**: Already tested in EDPIE project
5. **No new dependencies**: Uses existing sklearn

## Related Issues

- #7 - Chi-squared goodness-of-fit test for Mahalanobis
- #9 - Hierarchical clustering module (future)
- #10 - Clustering comparison tools (future)
- #11 - Cluster-aware outlier detection (future)

## Checklist

- [ ] Port `detect_outliers_kmeans()` from EDPIE
- [ ] Port `detect_outliers_gmm()` from EDPIE
- [ ] Port `create_kmeans_outlier_plots()` from EDPIE
- [ ] Port `create_gmm_outlier_plots()` from EDPIE
- [ ] Update `combine_outlier_methods()` to support new methods
- [ ] Write unit tests for K-Means detection
- [ ] Write unit tests for GMM detection
- [ ] Write visualization tests
- [ ] Test on real dataset (wheat phenotyping data)
- [ ] Update docstrings to match project style
- [ ] Update CHANGELOG.md
- [ ] Add example to documentation/README

**Comment by eberrigan** [comment]
Closed by PR #6 (Add comprehensive cross-experiment correlation analysis module).

The clustering-based outlier detection methods have been implemented:
- ✅ K-Means outlier detection: `detect_outliers_kmeans()` in outlier_detection.py
- ✅ GMM outlier detection: `detect_outliers_gmm()` in outlier_detection.py
- ✅ Core clustering module: clustering.py with `perform_kmeans_clustering()` and `perform_gmm_clustering()`
- ✅ Visualization functions in outlier_visualization.py and cluster_visualization.py
- ✅ Comprehensive test coverage

This issue has been fully implemented with the layered architecture described in issue #9.
